ci/steps/publish-npm.sh

```diff
@@ -47,22 +47,40 @@ main() {
     exit 1
   fi
 
-  # We need TAG to know what to publish under on npm
-  # Options are "latest", "beta", or "<pr number >"
-  # See Environment comments above to know when each is used.
-  if ! is_env_var_set "NPM_TAG"; then
-    echo "NPM_TAG is not set. This is needed for tagging the npm release."
+  # We use this to grab the PR_NUMBER
+  if ! is_env_var_set "GITHUB_REF"; then
```

Comment:
I didn't think about that but that's a great idea! I'm all for reducing complexity. Great suggestion! I will do that.

Reply:
Hmm...I thought threw refactoring it and there's a problem. If _only_ use `GITHUB_` environment variables to determine where to publish, we can't easily distinguish between `staging` and `production` since they both rely on `$GITHUB_EVENT_NAME == 'push' && $GITHUB_REF == 'refs/heads/master'`

What I can do though is move this if/else logic to the script and say, if `NPM_ENVIROMENT` isn't set, run this if/else logic. That cleans it up a bit. 

But yeah, the main problem is we can't distinguish between `ci.yaml` running and `npm-brew.yaml` unless we use something like [`GITHUB_WORKFLOW`](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables) and switch on the name. But if we change the name, then our script breaks. That feels more flakey than using something explicit like `NPM_ENVIRONMENT` IMO.

Reply:
Some of the time, right? I think it would also run on other events like `workflow_dispatch` and possibly any other type of `release` events (like a `prerelease`) I need to double check the names

<img width="875" alt="image" src="https://user-images.githubusercontent.com/3806031/152421794-9e2c227e-ab03-4edf-8213-1b6739b5ecf0.png">

Reply:
Seems like it would be:
- `workflow_dispatch`
- `release`

I think that's it actually.

Reply:
I mean I guess there are three scnearios:

- pushes into `main` only happen on PR merges. so that's `staging`
- pushes to a branch that are targeting `main` are `development`
- `workflow_dispatch` and `release` are releases so that's `production`

<img width="1095" alt="image" src="https://user-images.githubusercontent.com/3806031/152422360-b26c200b-3f99-4664-8c45-c217e9d24aab.png">

That could work potentially. Is this what you're hoping for? all the logic to live here in an `if/elseif/else`?

Reply:
Ahaha yeah I was just thinking about that.

We would have to check for the `workflow_dispatch` event...but it is possible we could add that to other workflows in the future so probably not a good idea.

Reply:
You know...I am not really sure what I was hoping for.  It felt like we had a mix of telling the script what to do but the script also having to figure out what to do so I think I wanted it all one way or another?

Reply:
Oh yeah, that's true! Hmm...well I think the current approach is solid then,  what do you think?

Reply:
Like do `yarn publish:npm` and have the script figure it all out or do `yarn publish:npm "$PR-GITHUB_SHA"`.

Probably not worth spinning our wheels over. :stuck_out_tongue:

Reply:
Current implementation works for me!

Reply:
Lol I feel ya. I wish GitHub environment variables made it easier. This shouldn't need such a convoluted solution 😂 Just tell me easily if this is a PR, a merge into main or a release...that's all i want. not asking for much

Reply:
hahaha yeah definitely

One downside of having the script figure out what to do is that you cannot really run this locally unless you set all the right variables although idk why you would want to.

Reply:
Agreed! Definitely a limitation of the current implementation. If we do run into that use case down the road, I will take responsibility for adjusting it so that you can run it locally!